libblake3: fix FreeBSD cross-compile builds

silvanshade opened the pull request:
This PR fixes `libblake3` cross-compile builds for FreeBSD.

It also patches `tbb` to fix the tests on FreeBSD and Windows.

Follows from discussion at https://github.com/NixOS/nixpkgs/pull/401156

- Built on platform(s)
  - [x] mingw32
  - [x] x86_64-freebsd
  - [x] x86_64-linux

silvanshade committed:
tbb: fix tests on FreeBSD and Windows

This adds @Ericson2314's upstream patch to the package.

silvanshade committed:
libblake3: fix FreeBSD cross-compile builds

Ericson2314 commented on pkgs/development/libraries/tbb/default.nix:
```diff
@@ -44,11 +44,11 @@ stdenv.mkDerivation rec {
-  cmakeFlags = [
-    # Skip tests to work around https://github.com/uxlfoundation/oneTBB/issues/1695
-    (lib.cmakeBool "TBB_TEST" (!stdenv.hostPlatform.isWindows))
+    (fetchpatch {
+      name = "fix-tbb-freebsd-and-windows-tests.patch";
+      url = "https://patch-diff.githubusercontent.com/raw/uxlfoundation/oneTBB/pull/1696.patch";
+      hash = "sha256-yjX2FkOK8bz29a/XSA7qXgQw9lxzx8VIgEBREW32NN4=";
+    })
```
Can you do this to the newer TBB too?

Ericson2314 commented on pkgs/by-name/li/libblake3/package.nix:
```diff
@@ -28,6 +29,20 @@ stdenv.mkDerivation (finalAttrs: {
+  ];
+
+  patchPhase = ''
+    pushd ..
+    patchPhase
+    popd
```
haha nice :)

fgaz replied:
Yes :)

However, `patch` has builtin support for this, so I suggest replacing it with `patchFlags = [ "-d" ".." ];` so we don't have to mess with phases

silvanshade commented:
I made some changes addressing the reviews, notably adding the `tbb` patch for the other newer `tbb` releases and using `patchFlags` instead of overriding `patchPhase`.

I also opted to include the BLAKE3 compiler detection patch as a local file rather than pulling from the upstream PR since upstream may want to include that as part of more significant CMake refactoring so it seems more stable to just host the change locally until that PR is resolved.

Ericson2314 commented:
Oh combination of this and #402739 breaks eval, because two `patches` lists. We're fixing.
